File: Non-procedural (functional) programming/Solved/antidvorak/male/batoh.pl
?- consult('../functional.pl').

%klasika - problem batohu

cmpare(B, Acc) :-
    map(snd, B, C),
    sum(C, S),
    map(snd, Acc, A),
    sum(A, S2),
    S >= S2.

batoh(W, Xs, Ys) :- 
    vsechnyBatohy(W, Xs, [], Yss),
    maximumBy(cmpare, Yss, Ys).

batoh(_, [], Acc, Acc).

batoh(W, [V-C | Xs], Acc, Out) :-
    T is W - V,
    T >= 0, !,
    ( batoh(T, Xs, [V-C|Acc], Out) ; batoh(W, Xs, Acc, Out) ).

batoh(W, [_ | Xs], Acc, Out) :- batoh(W, Xs, Acc, Out).

nanelezenyBatoh(W, Xs, Acc, Ys) :-
    batoh(W, Xs, [], Ys),
    \+ member(Ys, Acc).

vsechnyBatohy(W, Xs, Acc, Out) :-
    nanelezenyBatoh(W, Xs, Acc, Ys), !,
    vsechnyBatohy(W, Xs, [Ys|Acc], Out).

vsechnyBatohy(_, _, Acc, Acc).

File: Non-procedural Programming/Solved/antihric/male/batoh.hs
import Kombinace
import Functional

--klasika - problem batohu

batoh :: Int -> [(Int, Int)] -> [(Int, Int)]
batoh 0 _ = []
batoh w [] | w >= 0 = []
batoh w ((v,c) : ps) | c > 0 && vb + v <= w     = (v, c) : b
                     | c > 0 && vb + v > w      = maximumBy (\b b' -> cenaBatohu b > cenaBatohu b') noveBatohy
                     | c < 0                    = b
    where
        b = batoh w ps
        vb = vahaBatohu b
        nejsou = deleteFirstOccurence ((v,c):ps) b --nejsou v batohu
        noveBatohy = b : [nb | h <- [1..length b], k <- kombinace h b, j <- [1..length nejsou], k2 <- kombinace j nejsou, let nb = nejsou ++ (deleteFirstOccurence b k), vahaBatohu nb <= w]

vahaBatohu :: [(Int, Int)] -> Int
vahaBatohu = sum . map fst 

cenaBatohu :: [(Int, Int)] -> Int
cenaBatohu = sum . map snd

deleteFirstOccurence :: Eq a => [a] -> [a] -> [a]
deleteFirstOccurence xs [] = xs
deleteFirstOccurence xs (y : ys) = case i of 
                                        Just idx -> deleteFirstOccurence (deleteAt idx xs) ys
                                        Nothing  -> deleteFirstOccurence xs ys
            where
                i = elemIndex y xs